lib/Bagger/Agent/LW.pm

```diff
@@ -76,49 +81,91 @@ _recv() function.
 our %_INJECTION = ();
 
 sub _run_injection {
-    my ($injectpoint, @args) = @_;
-    return $_INJECTION{$injectpoint}(@args) if ref $_INJECTION{$injectpoint};
+    my $injectpoint = shift @_;
+    $_INJECTION{$injectpoint}(@_) if ref $_INJECTION{$injectpoint};
 }
 
+my $done = AnyEvent->condvar;
 my $recv;
 
 sub recv { $recv }
 
-sub run {
-
+sub start {
+    my $inject = $ENV{TEST_AGENT}; # faster
     # It is fine to die here if not found
     my $kvstoremod = Bagger::Storage::Config->get('kvstore_type')->value_string;
     my $kvstoreconfig = Bagger::Storage::Config->get('kvstore_config')->value;
    
-    my $kvstore = Bagger::Agent::KVstore->new($kvstoremod, $kvstoreconfig);
+    my $kvstore = Bagger::Agent::KVStore->new(module => $kvstoremod, config => $kvstoreconfig);
 
     # the callback is a coderef here for procesing the messages and sending them forward.
     # The bulk of the work here is in this function.
     
-    my $callback = sub {
+    my $callback = unblock_sub {
```

Comment:
This moves the callback into a non-event-loop coroutine (and possibly different thread) and therefore makes this call asynchronous.  That prevents the etcd driver's use of blocking for events from causing the main event loop from blocking.